cmd/bench/bench.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"code.cloudfoundry.org/bytefmt"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"github.com/olekukonko/tablewriter"
+	"github.com/spacemeshos/post/initialization"
+	"github.com/spacemeshos/post/proving"
+	"github.com/spacemeshos/post/shared"
+	"github.com/spacemeshos/post/validation"
+	"log"
+	"math"
+	"os"
+	"runtime"
+	"strconv"
+	"time"
+)
+
+type Config = shared.Config
+
+var (
+	id, _         = hex.DecodeString("deadbeef")
+	challenge, _  = hex.DecodeString("this is a challenge")
+	defaultConfig = shared.DefaultConfig()
+)
+
+func main() {
+	datadir := flag.String("datadir", defaultConfig.DataDir, "filesystem datadir path")
+	space := flag.Uint64("space", 1<<23, "space per unit, in bytes")
+	single := flag.Bool("single", false, "whether to execute a single test instead of the complete set")
+	flag.Parse()
+
+	log.Printf("bench config: datadir: %v, space: %v", *datadir, bytefmt.ByteSize(*space))
+
+	cases := genTestCases(*datadir, *space, *single)
```

Comment:
Consider allowing users to provide the cases manually. This would allow testing different specific cases quickly without changing code.

Pro-level would be to expose a way to generate a file with the test cases that `genTestCases` generates that the user can tweak and then run.

After typing this comment I almost didn't post it because I'm not sure it's worth the effort, but decided add this disclaimer so you can decide for yourself.

Reply:
For now I thought to have 3 modes: single, mid, full.
Single - already supported, but need to add more flags to fully control the config.
Full - some config args are auto-generated, using `genTestCases`.
Mid - same as Full, but only the critical tests.

Some flags will be redundant when not in `single` mode. 

What do you say?

Reply:
If you're gonna work on this I think my approach can be more flexible to use and easier to code.

Just expose `genTestCases` to the outside and make it return some `JSON` (or similar human-readable format). Then make the benchmark accept this same format. Doing so will allow users to manually control what runs and making changes to `genTestCases` will also be easy when it's separately exposed, imo.

But... up to you... this is just how I imaging I'd like it to work, as a user.

Reply:
I wanted to avoid the need for a smart user, and stick with simple the cmd args. 
So if the user wants control, he can use the `single` mode, and run it many times.
Otherwise, `genTestCases` will create the same important tests that we can benchmark periodically.

But i’ll keep that in mind and add it if needed.